internal/docstore/driver/codec.go

```diff
@@ -0,0 +1,701 @@
+// the resulting byte slice. Here proto is the package "github.com/golang/protobuf/proto".
+//
+// Not every map key type can be encoded. Only strings, integers (signed or
+// unsigned), and types that implement encoding.TextMarshaler are permitted as map
+// keys. These restrictions match exactly those of the encoding/json package.
+func Encode(v reflect.Value, e Encoder) error {
+	return wrap(encode(v, e), gcerr.InvalidArgument)
+}
+
+func encode(v reflect.Value, enc Encoder) error {
+	if !v.IsValid() {
+		enc.EncodeNull()
+		return nil
+	}
+	if v.Type().Implements(binaryMarshalerType) {
+		bytes, err := v.Interface().(encoding.BinaryMarshaler).MarshalBinary()
+		if err != nil {
+			return err
+		}
+		enc.EncodeBytes(bytes)
+		return nil
+	}
+	if v.Type().Implements(protoMessageType) {
+		if v.IsNil() {
+			enc.EncodeNull()
+		} else {
+			bytes, err := proto.Marshal(v.Interface().(proto.Message))
+			if err != nil {
+				return err
+			}
+			enc.EncodeBytes(bytes)
+		}
+		return nil
+	}
+	if v.Type().Implements(textMarshalerType) {
+		bytes, err := v.Interface().(encoding.TextMarshaler).MarshalText()
+		if err != nil {
+			return err
+		}
+		enc.EncodeString(string(bytes))
+		return nil
+	}
+	switch v.Kind() {
+	case reflect.Bool:
+		enc.EncodeBool(v.Bool())
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		enc.EncodeInt(v.Int())
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		enc.EncodeUint(v.Uint())
+	case reflect.Float32, reflect.Float64:
+		enc.EncodeFloat(v.Float())
+	case reflect.Complex64, reflect.Complex128:
+		enc.EncodeComplex(v.Complex())
+	case reflect.String:
+		enc.EncodeString(v.String())
+	case reflect.Slice:
+		if v.IsNil() {
+			enc.EncodeNull()
+			return nil
+		}
+		fallthrough
+	case reflect.Array:
+		return encodeList(v, enc)
+	case reflect.Map:
+		return encodeMap(v, enc)
+	case reflect.Ptr:
+		if v.IsNil() {
+			enc.EncodeNull()
+			return nil
+		}
+		return encode(v.Elem(), enc)
+	case reflect.Interface:
+		if v.IsNil() {
+			enc.EncodeNull()
+			return nil
+		}
+		return encode(v.Elem(), enc)
+
+	case reflect.Struct:
+		return encodeStruct(v, enc)
+
+	default:
+		return gcerr.Newf(gcerr.InvalidArgument, nil, "cannot encode type %s", v.Type())
+	}
+	return nil
+}
+
+// Encode an array or non-nil slice.
+func encodeList(v reflect.Value, enc Encoder) error {
+	// Byte slices or byte arrays encode specially.
+	if v.Type().Elem().Kind() == reflect.Uint8 {
+		enc.EncodeBytes(v.Bytes())
+		return nil
+	}
+	n := v.Len()
+	enc2 := enc.EncodeList(n)
+	for i := 0; i < n; i++ {
+		if err := encode(v.Index(i), enc2); err != nil {
+			return err
+		}
+		enc2.ListIndex(i)
+	}
+	return nil
+}
+
+// Encode a map.
+func encodeMap(v reflect.Value, enc Encoder) error {
+	if v.IsNil() {
+		enc.EncodeNull()
+		return nil
+	}
+	keys := v.MapKeys()
+	enc2 := enc.EncodeMap(len(keys))
+	for _, k := range keys {
```

Comment:
Isn't there a way to do `range k,v`?

Reply:
Yes, `reflect.MapRange`. But it's new in Go 1.12, so we can't use it yet.